HP-318 step 6: add and edit emails, phones and addresses

NikoHelle opened the pull request:
Added components for viewing and editing emails, phones and addresses. 

Old data can be saved and new added and saved, but "remove"-button is not added for deleting items.

If there is a need to remove data, it can be done in https://profiili.dev.hel.ninja/

NikoHelle committed:
Add and remove items added to EditData.ts

New items can be added and removed. There is no button yet in the UI for removing old data. New item is removed when cancel is pressed.

Added id to createFormValues(), otherwise all items are always removed and re-created when saving.

NikoHelle committed:
Add and remove item to useProfileDataEditor

Only new item removal is supported in this commit

NikoHelle committed:
Common error messaging added to BasicData.tsx

All components need same hasFieldError and getFieldErrorMessage -functions, so added them to formik.ts

NikoHelle committed:
Check invalid values in BasicData.test.tsx

This also required update to componentMocking.tsx

ClickElement returns HTMLElement that was pressed. Element's "disabled"-property needs sometimes checking.

submit() has new parameter: optionalSubmitButtonSelector, because email, phones and addresses components will introduce different submit button selectors.

new functions for element attribute checking.

waitForElementAndValue() moved above its usage in setInputValue()

NikoHelle committed:
Render and edit addresses, emails and phones

Components for handling multiItemProfileData.

MultiItemEditor handles rendering of item lists  and also handles common actions like saving and renders save indicators. It also shows button for adding new items.

MultiItemRow renders emails and phones, because those have same number of fields and are simple to handle.

MultiItemAddressRow is almost identical but different enough to have its own component for handling addresses

MultiItem*Row are like BasicData.tsx, but also handle new items.

useCommonEditHandling is used by MultiItem*Row because they have identical edit functions.

Remove and set-primary -buttons are not rendered at the moment.

NikoHelle committed:
Update editData updateItems test for all dataTypes

NikoHelle committed:
Fix bug in useProfileMutations.test.ts

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -60,11 +60,12 @@ export type TestTools = RenderResult & {
     selector: ElementSelector
```
File comment: This file is quite large. It seems to contain test related helper functions and types? I just wonder if it's doing too much, as the name componentMocking feels like it does not match for all what the file exposes. For example functions getElementAttribute and waitForElementAttributeValue. They feel like helper functions for testing, not specifically about mocking components? Not sure about how it could be improved though, just an observation 😅

NikoHelle replied:
Yep. 👍   

Could split types, helpers and rendering?

NikoHelle replied:
Done.

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -86,6 +87,30 @@ export const cleanComponentMocks = (): void => {
+export const submitButtonSelector: ElementSelector = {
+  querySelector: 'button[type="submit"]',
+};
+export const getElementAttribute = (
```
Not sure if this function brings extra value? Could we just use element.getAttribute directly?

NikoHelle replied:
Yes, there used to be check for element existence in the func, but not anymore. I'll remove the func.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/common/test/componentMocking.tsx:
```diff
@@ -231,27 +256,29 @@ export const renderComponentWithMocksAndContexts = async (
+    const { selector, value } = props;
+    return waitFor(async () => {
+      const elementValue = await getTextOrInputValue(selector);
+      if (value && (!elementValue || !elementValue.includes(value))) {
```
What if `value` is false in this condition? It seems to wait for elementValue?

NikoHelle replied:
Value cannot be false, it must be a string. "Value" is inside the "if" to cover the case when it is an empty string ''.

This is quite unclear and I needed some time to find out why it is there. I changed that part to be a lot clearer:

if (value === '' && elementValue === '') {
        return;
      }

ElementValue can be "undefined" until element is found in the dom. That's why !elementValue wont work.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/components/basicData/__tests__/BasicData.test.tsx:
```diff
@@ -181,4 +184,75 @@ describe('<BasicData /> ', () => {
+          'aria-invalid',
+          'true'
+        );
+        // getElement throws is element is not found
```
typo: is -> if

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/components/basicData/__tests__/BasicData.test.tsx:
```diff
@@ -181,4 +184,75 @@ describe('<BasicData /> ', () => {
+        );
+        // getElement throws is element is not found
+        // so element is not checked in expect
+        errorElementGetter();
```
This probably works already as it is, but could we use this here in addition:

`expect(() => ...error...).not.toThrow(error)`?

NikoHelle replied:
This is probably clearer

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/hooks/useProfileDataEditor.ts:
```diff
@@ -96,14 +102,31 @@ export function useProfileDataEditor({
     save,
     reset,
+    add,
+    hasNew,
```
Could we inline the hasNewItem here?

NikoHelle replied:
Do you mean adding 

`hasNew: ()=>hasNewItem?`
 
It is not allowed by sonarCube, I think. It wont allow mixing shorthands with other code in "return {...}" blocks.

But I changed the code where hasNewItem is created to 

`hasNewItem: hasNew,`

so "const hasNew = () => hasNewItem();"  is removed.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/helpers/formik.ts:
```diff
@@ -51,3 +51,24 @@ export function getFieldError<FormValues>(
 }
+
+export function createFormFieldHelpers<FormikValues>(
+  t: TFunction,
```
In general I think variables with only 1 letter should be avoided, since they are not very descriptive and they are more harder to find with text editors (for example ctrl + f). IDEs can help though. I guess this is the translate function (t) of the i18next library? Probably not much then what could be done here, as it might be confusing to use another name for it.

NikoHelle replied:
Yes, "t" it is **the** i18n.t, so kept the same name with the function. One letter funcs/vars are bad, but in this case it clarifies that it is **the** "t" -function.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -115,11 +118,15 @@ function isSaving(allItems: EditData[]): boolean {
 }
 
-function hasNewItem(allItems: EditData[]): boolean {
+export function getNewItem(allItems: EditData[]): EditData | undefined {
```
Is there always only 1 new item in the array? Could it be more safe to rename this: getFirstNewItem?

NikoHelle replied:
There can be only one (or none), so no need to clarify it.

VilleMiekkoja commented on src/profile/helpers/editData.ts:
```diff
@@ -103,9 +103,12 @@ export type EditFunctions = {
   resetItem: (targetItem: EditData) => boolean;
+  addItem: () => EditData;
+  hasNewItem: () => boolean;
+  removeItem: (targetItem: EditData) => Partial<FormValues> | null;
```
Why do we wish to sometimes return partial FormValues and sometimes return null?

NikoHelle replied:
When formValues are returned the caller knows it needs to save something. Null is returned when a new, unsaved, item is removed and there is no need for server call.

VilleMiekkoja commented on src/profile/components/multiItemRow/multiItemRow.module.css:
```diff
@@ -0,0 +1,43 @@
+  flex-direction: column;
+}
+
+/* fixes bug in HDS TextInput when it is wrapped to an element with display:flex: */
```
There might be a fix coming to this in the next release of HDS 😅 . We removed TextInput input wrapper height 100%, which might be related to this.

VilleMiekkoja replied:
The fix should now be available: https://www.npmjs.com/package/hds-react

NikoHelle replied:
NumberInput, PhoneInput and other HDS fixes mentioned here were released after this was originally created. HDS update would probably cause a lot of changes, so did not want to update it in this PR and make it larger. I'd rather make a new ticket for HDS updates.

NikoHelle replied:
HDS update ticket created https://helsinkisolutionoffice.atlassian.net/browse/HP-727

VilleMiekkoja commented on src/profile/components/multiItemRow/multiItemRow.module.css:
```diff
@@ -0,0 +1,43 @@
+  width: 100%;
+}
+
+.editableRow {
```
Airbnb styleguide prefers dashes over camelCasing in class names: https://github.com/airbnb/css#css

dev.hel.fi recommends following the airbnb's style guide with some exceptions: https://dev.hel.fi/coding-standards#reactjsx-style-guide

Doesn't refer to css there though, so not sure if it should be followed in styles also?

NikoHelle replied:
I have kept the same naming that all other css files follow in the Profile UI. Original files are probably older than the spec in dev.hel.fi. Maybe all files should be changed in one update or just this one now?

Maybe combine this to a HDS update.

VilleMiekkoja replied:
Yeah it might be a big task to change them all, so not sure is it worth it. Though if we're still in the beginning part of creating the open-city-profile-ui, it might be worth it.

NikoHelle replied:
There wont be any new css files :)

VilleMiekkoja commented on src/profile/components/multiItemRow/multiItemRow.module.css:
```diff
@@ -0,0 +1,43 @@
+}
+
+.value {
+  font-size: 18px;
```
Could we apply the hds design tokens here? https://hds.hel.fi/design-tokens/typography#tokens. I think it would be the `$fontsize-body-l: 1.125rem; // 18px`

NikoHelle replied:
Yes, HDS should be followed. Another fix to be done in the HDS update....

VilleMiekkoja commented on src/profile/hooks/useCommonEditHandling.ts:
```diff
@@ -0,0 +1,53 @@
+): UseActionHandlingReturnType => {
+  const { data, onAction } = props;
+  const isNew = isNewItem(data);
+  const [isEditing, setEditing] = useState(isNew);
```
isNew means it is isEditing? Asking because it is passed as the default value here

NikoHelle replied:
No, if the item is a new one, then edit mode should always be "active".

VilleMiekkoja commented on src/profile/hooks/useCommonEditHandling.ts:
```diff
@@ -0,0 +1,53 @@
+    const [err] = await to(onAction(action, data, newValue as EditDataValue));
+    if (isNew) {
+      if (err) {
+        setCurrentAction(undefined);
```
Should we also reject on error? Since there's no return inside this if, the actionHandler will resolve on line 32 even in the case of error?

NikoHelle replied:
Awaiting function does not care what happens, because all error handling is here. Adding rejection would also require rejection handling to the awaiting function, because all promise rejections must be handled.

VilleMiekkoja commented on src/profile/components/multiItemRow/MultiItemRow.tsx:
```diff
@@ -0,0 +1,122 @@
+  const schema = dataType === 'phones' ? phoneSchema : emailSchema;
+  const inputValue: string =
+    (dataType === 'phones'
+      ? (value as PhoneValue).phone
```
It's little hard to follow quickly where value comes from. Would maybe prefer applying full path like: `props.data.value` that shows it right away that it comes from props. Or then try to destructure it only once from props.

NikoHelle replied:
Doesn't this apply to all destructured variables? So never use it? I'm a bit lost how to solve this and make it clearer:
```
const { data, testId, dataType } = props;
...
const { value, primary } = data;
```

VilleMiekkoja replied:
Yeah it probably does 😅 . I just think that if we get a variable by destructuring multiple times at different rows, it makes it little harder to read where it originally came from. Is it possible to get it by destructuring it only once? Or then just apply the props.data.value directly

NikoHelle replied:
Changed to const { data:{ value , primary }, testId, dataType } = props;

VilleMiekkoja commented on src/profile/components/multiItemRow/MultiItemRow.tsx:
```diff
@@ -0,0 +1,122 @@
+                  name="value"
+                  id={inputId}
+                  maxLength={formFields.value.max as number}
+                  as={TextInput}
```
Should we apply the HDS PhoneInput component here when type is phone? https://hds.hel.fi/components/phone-input

One benefit is that it adds the type=tel to the input, which then helps mobile users by opening the number keyboard

NikoHelle replied:
NumberInput, PhoneInput and other HDS fixes mentioned here were released after this was originally created. HDS update would probably cause a lot of changes, so did not want to update it in this PR and make it larger. I'd rather make a new ticket for HDS updates.

VilleMiekkoja commented on src/profile/components/multiItemRow/MultiItemRow.tsx:
```diff
@@ -0,0 +1,122 @@
+
+type EmailAndPhoneFormikValue = { value: string };
+
+function MultiItemRow(props: RowItemProps): React.ReactElement {
```
I fear there can be a risk that combining phone and email to same row component might not scale in the long run in terms of readability of the code. It makes code more abstractly named, and there's going to be places where you have to check the data-type, which increases the complexity.

NikoHelle replied:
IMHO, data-type checks are easier to follow than making two files (EmailRow and PhoneRow) and then more files for common code of those two. The need for scaling this file is an abstract scenario. If there are new features, it is easier to add them here?

Currently there are 3 places where dataType is checked and 2 where it is passed as argument.
There are 122 lines of code and only those 3 lines would not be common if split in to multiple files.

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+                name="postalCode"
+                id={`${testId}-postalCode`}
+                maxLength={formFields.postalCode.max as number}
+                as={TextInput}
```
Could we apply HDS NumberInput component here? https://hds.hel.fi/components/number-input

It aids screen readers and gives better mobile keyboard

NikoHelle replied:
NumberInput, PhoneInput and other HDS fixes mentioned here were released after this was originally created. HDS update would probably cause a lot of changes, so did not want to update it in this PR and make it larger. I'd rather make a new ticket for HDS updates.

VilleMiekkoja commented on src/profile/helpers/__tests__/editData.test.ts:
```diff
@@ -98,6 +102,32 @@ describe('editData.ts ', () => {
+    dataType: EditDataType
+  ): MultiItemProfileNode[] =>
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    //@ts-ignore
```
I wonder if there's a way to type the row 129 without typescript warnings?

NikoHelle replied:
Yes, if changed to this:

```
const pickMultiItemFormValues = (
    formValues: FormValues | Partial<FormValues>,
    dataType: EditDataType
  ): MultiItemProfileNode[] => {
    if (
      dataType === 'addresses' ||
      dataType === 'emails' ||
      dataType === 'phones'
    ) {
      return formValues[dataType] as MultiItemProfileNode[];
    }
    throw new Error('Invalid datatype for picking MultiItemProfileNodes');
  };
```

I liked the previous version because this is a test file.

NikoHelle replied:
Done.

VilleMiekkoja commented on src/profile/helpers/__tests__/editData.test.ts:
```diff
@@ -172,16 +202,16 @@ describe('editData.ts ', () => {
         updateItemAndCreateSaveData,
```
Off-topic: comment for line 193: my IDE suggest `!!node.primary` could be simplified to `node.primary`

NikoHelle replied:
changed. Required a new commit, because this wasn't added in this PR

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+  const { t, i18n } = useTranslation();
+  const lang = i18n.languages[0];
+  const applicationLanguage = getLanguageCode(i18n.languages[0]);
+  const countryList = countries.getNames(applicationLanguage);
```
Would it be good idea to sort the countries based on names? Most of them seemed to be sorted already, but I found in the end of the list there's countries that are not sorted with the rest of the countries?
![image](https://user-images.githubusercontent.com/2777633/116708223-2a248a80-a9d8-11eb-9d51-b1c96aa92f8c.png)

NikoHelle replied:
What language is your browser or profile? I've different ordering.

NikoHelle replied:
If UI language is English, the sorting is messed up. In Finnish version, only "Kosovo" is in wrong place. Ill' sort them manually.

NikoHelle replied:
Added sorting

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+              <Field
+                name="postalCode"
+                id={`${testId}-postalCode`}
+                maxLength={formFields.postalCode.max as number}
```
Tested this and the maxLength seemed to be 5? I think there might be longer postal codes in other countries, so should we support longer ones as well, because we want to support other countries too?

NikoHelle replied:
There is no documentation why 5 was selected in old Profile UI. Perhaps, because BE didn't always support more chars. I asked around and now BE supports 32 chars. UX designer approved this too. So will change to 32.

NikoHelle replied:
Done

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+              />
+              <div className={commonFormStyles.formField}>
+                <FormikDropdown
+                  name="countryCode"
```
If the country is USA, and we wish to support USA addresses, should we add state field also, when USA is selected? Im not sure if state is compulsory for USA address? 😅

NikoHelle replied:
Good point. Currently BE has no support for US states or whatever China or Zimbabwe needs 😃  This needs additional designing.

VilleMiekkoja commented on src/profile/helpers/__tests__/editData.test.ts:
```diff
@@ -254,6 +288,234 @@ describe('editData.ts ', () => {
+          expect((newItem.value as AddressValue).address).toBeDefined();
+        }
+      });
+      it(`updateItemAndCreateSaveData updates the new ${dataType} item and it is in returned formValues`, () => {
```
I wonder if the language of this sentence could be improved? Im not sure I got the ending part of it?

NikoHelle replied:
Fixed.

VilleMiekkoja commented on src/profile/helpers/__tests__/editData.test.ts:
```diff
@@ -254,6 +288,234 @@ describe('editData.ts ', () => {
+        expect(itemsInFormValues[testIndex].id).toEqual(
+          editDataWithRemovedItem[testIndex + 1].id
+        );
+        // edit datas have not changed yet
```
edit datas -> edit data

NikoHelle replied:
Fixed.

VilleMiekkoja commented on src/profile/components/multiItemEditor/MultiItemEditor.tsx:
```diff
@@ -0,0 +1,173 @@
+  );
+}
+
+export default MultiItemEditor;
```
Should we use the inline export instead of this?

NikoHelle replied:
Following other exports in .tsx-files here

VilleMiekkoja replied:
Ok, sorry I thought that inline export was the convention.

VilleMiekkoja commented on src/profile/components/multiItemEditor/MultiItemEditor.tsx:
```diff
@@ -0,0 +1,173 @@
+  const isAddButtonDisabled = hasNew();
+  const RowComponent = hasAddressList ? MultiItemAddressRow : MultiItemRow;
+  const texts = (function() {
+    if (dataType === 'emails') {
```
Hmm.. I wonder if there's a risk that this file becomes too complex, since we are trying to support all different dataTypes? If the requirements of different data types start changing a lot, it might be a good idea to apply some refactoring.

NikoHelle replied:
Actually supporting only 3 data types. :) Those who have multiple items and have same editing tools and same notifications.

All this does it handle identical actions (save, add ,remove, set-primary), show notifications and render each item.

NikoHelle replied:
Multiple items need a "parent / wrapper" that knows what children are doing. Later on, it must prevent children from being edited while re-ordering is in progress. Now it prevents multiple new items. 

"If the requirements of different data types start changing a lot, "

Then it should be addressed, perhaps not now?

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+  );
+}
+
+export default MultiItemAddressRow;
```
Should we change this export to inline?

NikoHelle replied:
Following other exports in .tsx-files here.

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+import { TextInput } from 'hds-react';
```
Looking good! I wonder if the name of the file could be simplified to: AddressRow?

NikoHelle replied:
I chose "MultiItemAddressRow" so it would group with "MultiItemRow" and "MultiItemEditor"

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+                aria-labelledby={`${dataType}-city-helper`}
+              />
+              <div className={commonFormStyles.formField}>
+                <FormikDropdown
```
EDIT: was about to suggest applying HDS Dropdown, but then noticed that the FormikDropdown applies it.

Combobox might fit better, since there's more than 8 options, and it might be quicker for the user to input first letter of the country and get the result?

NikoHelle replied:
Currently you can type a letter and it jumps there. I don't know what HDS combobox is and when it was introduced. The designs do not use it.

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+              />
+              <div className={commonFormStyles.formField}>
+                <FormikDropdown
+                  name="countryCode"
```
Should this be country instead of countryCode?

NikoHelle replied:
BE, profileData and all other places us countryCode.

VilleMiekkoja commented on src/profile/components/multiItemEditor/__tests__/MultiItemEditor.test.tsx:
```diff
@@ -0,0 +1,362 @@
+      }
+    };
+
+    describe(`Renders ${dataType} and fields of that 
```
I wonder if the language could be improved here hmm.. _Renders address type and its fields:_?

NikoHelle replied:
Yes, I'll translate it to English :)

NikoHelle replied:
Fixed.

VilleMiekkoja commented on src/profile/components/multiItemEditor/__tests__/MultiItemEditor.test.tsx:
```diff
@@ -0,0 +1,362 @@
+        });
+      });
+
+      it(`Component shows error notification and stays in edit mode when error occures. 
```
Typo: occures -> occurs

NikoHelle replied:
Fixed.

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+                invalid={hasFieldError(formikProps, 'address')}
+                aria-invalid={hasFieldError(formikProps, 'address')}
+                helperText={getFieldErrorMessage(formikProps, 'address')}
+                labelText={t(formFields.address.translationKey)}
```
The font-weight and size of these input labels seem to change depending on mode (edit or view). It might be what is wanted, but at first glimpse looked little weird.

NikoHelle replied:
Looks like HDS and layouts have different styles. HDS should be followed. Another fix to be done in the HDS update....

VilleMiekkoja commented on src/profile/components/multiItemAddressRow/MultiItemAddressRow.tsx:
```diff
@@ -0,0 +1,189 @@
+                ? t('profileInformation.primaryAddress')
+                : t('profileInformation.address')}
+            </h4>
+            <div className={commonFormStyles.multiItemWrapper}>
```
The email and phone rows have a gray background color, I wonder if we would want the address row to have it too?

![image](https://user-images.githubusercontent.com/2777633/116869394-55e28300-ac19-11eb-8082-a2d511dbf614.png)

NikoHelle replied:
That is not in the layouts, so probably not wanted.

VilleMiekkoja commented:
Tested this locally, and seems to be working great! Good job with this 🙌 .

NikoHelle committed:
EditData.test.ts minor boolean test change

NikoHelle committed:
Changed postalCode max length to 32

NikoHelle committed:
Split componentMocking.tsx into smaller files

componentMocking.tsx grew too much and tried to do much more than name suggests. Instead of importing componentMocking.tsx, now use testingLibraryTools.ts

React components are moved their own files
ExposeHooksForTesting.tsx has hook test utils
renderComponentTestDOM creates the DOM with mockProviders and contexts.

No new code added, just splitted and removed componentMocking.tsx.
